rama-tls-rustls/src/server/acceptor_data.rs

```diff
@@ -1,47 +1,113 @@
 use crate::dep::pki_types::{CertificateDer, PrivateKeyDer, PrivatePkcs8KeyDer};
 use crate::dep::rcgen::{self, KeyPair};
-use crate::dep::rustls::ServerConfig;
+use crate::dep::rustls;
 use crate::key_log::KeyLogFile;
 use rama_core::error::{ErrorContext, OpaqueError};
 use rama_net::address::{Domain, Host};
 use rama_net::tls::server::SelfSignedData;
 use rama_net::tls::{ApplicationProtocol, KeyLogIntent};
 use rustls::ALL_VERSIONS;
+use std::pin::Pin;
 use std::sync::Arc;
 
-#[derive(Debug, Clone)]
+#[derive(Clone, Debug)]
 /// Internal data used as configuration/input for the [`super::TlsAcceptorService`].
 ///
 /// Created by converting a [`rustls::ServerConfig`] into it directly,
 /// or by using [`TlsAcceptorDataBuilder`] to create this in a more ergonomic way.
 pub struct TlsAcceptorData {
-    pub(super) server_config: Arc<ServerConfig>,
+    pub(super) server_config: ServerConfig,
 }
 
-impl From<ServerConfig> for TlsAcceptorData {
+#[derive(Clone)]
+/// [`ServerConfig`] used to configure rustls
+///
+/// This can either be a directly stored [`rustls::ServerConfig`], or a [`rustls::ServerConfig`]
+/// returned by a [`DynamicConfigProvider`] based on the received client hello
+pub(super) enum ServerConfig {
+    Stored(Arc<rustls::ServerConfig>),
+    Async(Arc<dyn DynDynamicConfigProvider + Send + Sync>),
+}
+
+impl std::fmt::Debug for ServerConfig {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Stored(arg0) => f.debug_tuple("Stored").field(arg0).finish(),
+            Self::Async(_) => f
+                .debug_tuple("Async")
+                .field(&"dynamic config provider")
+                .finish(),
+        }
+    }
+}
+
+impl From<rustls::ServerConfig> for TlsAcceptorData {
     #[inline]
-    fn from(value: ServerConfig) -> Self {
+    fn from(value: rustls::ServerConfig) -> Self {
         Arc::new(value).into()
     }
 }
 
-impl From<Arc<ServerConfig>> for TlsAcceptorData {
-    fn from(value: Arc<ServerConfig>) -> Self {
+impl From<Arc<rustls::ServerConfig>> for TlsAcceptorData {
+    fn from(value: Arc<rustls::ServerConfig>) -> Self {
         Self {
-            server_config: value,
+            server_config: ServerConfig::Stored(value),
         }
     }
 }
 
+impl<D: DynamicConfigProvider> From<D> for TlsAcceptorData {
+    fn from(value: D) -> Self {
+        Arc::new(value).into()
+    }
+}
+
+impl<D: DynamicConfigProvider> From<Arc<D>> for TlsAcceptorData {
+    fn from(value: Arc<D>) -> Self {
+        Self {
+            server_config: ServerConfig::Async(value),
+        }
+    }
+}
+
+pub trait DynamicConfigProvider: Send + Sync + 'static {
```

Comment:
Implement it also already for an `FnOnce + Clone` please. This will also simplify your example as you can just make an `async fn` and pass that as dynamic provider

Reply:
Nvm won't work because of `ClientHello` being a reference in Rustls